Optional Delayed Allocation on Node leave

kimchy opened the pull request:
Allow to set delayed allocation timeout on unassigned shards when a node leaves the cluster. This allows to wait for the node to come back for a specific period in order to try and assign the shards back to it to reduce shards movements and unnecessary relocations.

The setting is an index level setting under `index.unassigned.node_left.delayed_timeout` and defaults to 0 (== no delayed allocation). We might want to change the default, but lets do it in a different change to come up with the best value for it. The setting can be updated dynamically.

When shards are delayed, a log message with "info" level will notify which shards are being delayed and for how long.

An implementation note, we really only need to care about delaying allocation on unassigned replica shards. If the primary shard is uniassigned, anyhow we are going to wait for a copy of it, so really the only case to delay allocation is for replicas.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java:
```diff
@@ -125,6 +135,47 @@ public void clusterChanged(ClusterChangedEvent event) {
+                logger.info("delaying unassigned shard allocation, shards: {}", sb);
+            }
         } else {
             FutureUtils.cancel(scheduledRoutingTableFuture);
```
don't we need to cancel all the futures in delayedShardsToReroute ?

kimchy replied:
I will refactor it where its not relevant

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java:
```diff
@@ -125,6 +135,47 @@ public void clusterChanged(ClusterChangedEvent event) {
+                }
+            }
+            if (scheduled) {
+                logger.info("delaying unassigned shard allocation, shards: {}", sb);
```
this can very verbose (600 shards on a node). I'm doubting whether we should have this as debug and have an info level log saying "allocation of [X] unassigned shards delayed".

kimchy replied:
I can go with just a number, sure

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java:
```diff
@@ -141,6 +146,32 @@ public String getDetails() {
+            return 0;
+        }
+        long delta = Math.max(0l, System.currentTimeMillis() - timestamp);
+        if (delta == 0) {
```
I think I see what you are doing here (protect against system clock going way back in time), but I wonder if we should be so strict.  0 is a dangerous value - I'm worried that a very fast machine will render this check ineffective, so I would go with `delta<0`. I also wonder if a short ntp correction will release pending shards. Maybe use `delta<-delayTimeout`? .

kimchy replied:
I am not sure I follow, we are conservative here, if its less than 0, we treat it as no delay (which is the meaning of 0)

kimchy replied:
I think I got what you mean, I will improve it to check on negative value and return 0, to accomdate cases where its fast enough where currentTime is the same as timestamp

bleskes replied:
++. Thx

On Wed, Jun 17, 2015 at 9:09 PM, Shay Banon <notifications@github.com>
wrote:

>> +            return 0;
>> +        }
>> +        TimeValue delayTimeout = indexSettings.getAsTime(DELAYED_NODE_LEFT_TIMEOUT, settings.getAsTime(DELAYED_NODE_LEFT_TIMEOUT, DEFAULT_DELAYED_NODE_LEFT_TIMEOUT));
>> +        return Math.max(0l, delayTimeout.millis());
>> +    }
>> +
>> +    /**
>> +     * The delay in millis when delaying assigning the shard need to expire in.
>> +     */
>> +    public long getDelayAllocationExpirationIn(Settings settings, Settings indexSettings) {
>> +        long delayTimeout = getAllocationDelayTimeoutSetting(settings, indexSettings);
>> +        if (delayTimeout == 0) {
>> +            return 0;
>> +        }
>> +        long delta = Math.max(0l, System.currentTimeMillis() - timestamp);
>> +        if (delta == 0) {
> I think I got what you mean, I will improve it to check on negative value and return 0, to accomdate cases where its fast enough where currentTime is the same as timestamp
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/elastic/elasticsearch/pull/11712/files#r32663140

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/cluster/health/TransportClusterHealthAction.java:
```diff
@@ -264,11 +267,29 @@ private ClusterHealthResponse clusterHealth(ClusterHealthRequest request, Cluste
+        return new ClusterHealthResponse(clusterName.value(), concreteIndices, clusterState, numberOfPendingTasks, numberOfInFlightFetch, calculateDelayedUnassigned(clusterState));
+    }
+
+    private int calculateDelayedUnassigned(ClusterState clusterState) {
```
can we fold this into ClusterHealthResponse? that way we can test this as well as part of the unit testing.

kimchy replied:
We can't do it in ClusterHealthResponse since we need Settings, but I can move it to ClusterState with a provided Settings.

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/DelayedAllocationTests.java:
```diff
@@ -0,0 +1,149 @@
+                .put(UnassignedInfo.DEFAULT_DELAYED_NODE_LEFT_TIMEOUT, 0)).get();
+        ensureGreen("test");
+        internalCluster().stopRandomNode(InternalTestCluster.nameFilter(findNodeWithShard()));
+        assertBusy(new Runnable() {
```
Isn't this just an ensure green?

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java:
```diff
@@ -253,4 +257,27 @@ public void testFailedShard() {
+        UnassignedInfo unassignedInfo = new UnassignedInfo(UnassignedInfo.Reason.NODE_LEFT, null);
+        long delay = unassignedInfo.getAllocationDelayTimeoutSetting(Settings.builder().put(UnassignedInfo.DELAYED_NODE_LEFT_TIMEOUT, "10h").build(), Settings.EMPTY);
+        assertThat(delay, equalTo(TimeValue.timeValueHours(10).millis()));
+        Thread.sleep(100);
```
can we change this to await busy? then we don't need sleep ...

kimchy replied:
we could, it might end up waiting more, but sure

bleskes commented:
Did a review cycle. I like how things come together. One concern I had was the implementation in RoutingService where we maintain a queue of pending reroutes per unassigned (delayed) shard. I think it  will be simpler to just use the single future we already have and set it every time to the next expected change moment (i.e., the minimum delay of all unassigned shards). On a setting change we can do a reroute all the time (which we might do already). Am I missing something?

bleskes commented:
one more little thing - we need do some docs work as well, it's an important change. I can help if need be.

kimchy commented:
I pushed another round, mainly simplifying the code, adding more unit tests, and addressing comments. @bleskes once we agree on this as the way forward, I will add docs

s1monw commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java:
```diff
@@ -125,6 +131,21 @@ public void clusterChanged(ClusterChangedEvent event) {
+                registeredNextDelaySetting = nextDelaySetting;
+                TimeValue nextDelay = TimeValue.timeValueMillis(UnassignedInfo.findNextDelayedAllocationIn(settings, event.state()));
+                logger.info("delaying allocation for [{}] unassigned shards, next check in [{}]", UnassignedInfo.getNumberOfDelayedUnassigned(settings, event.state()), nextDelay);
+                registeredNextDelayFuture = threadPool.schedule(nextDelay, ThreadPool.Names.SAME, new Runnable() {
```
can we use `AbstractRunnable` here?

kimchy replied:
sure

s1monw commented on core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java:
```diff
@@ -141,6 +148,101 @@ public String getDetails() {
+            if (shard.primary()) {
+                continue;
+            }
+            if (shard.unassignedInfo() == null) {
```
wait I thought it can't be null?

kimchy replied:
yea, I will change to not worry about it

s1monw commented on core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java:
```diff
@@ -141,6 +148,101 @@ public String getDetails() {
+    public static long findNextDelayedAllocationSetting(Settings settings, ClusterState state) {
+        long nextDelaySetting = Long.MAX_VALUE;
+        for (ShardRouting shard : state.routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)) {
+            if (shard.primary()) {
```
instead of using `continue` can we invert the conditionals?

kimchy replied:
ok

kimchy commented:
@s1monw applied another round of changes

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java:
```diff
@@ -141,6 +148,89 @@ public String getDetails() {
+    }
+
+    /**
+     * The delay in millis when delaying assigning the shard need to expire in.
```
Got confused by this and had to go to the code :) - I think this will be clearer "returns the time in millisecond until the an unassigned shard can be reassigned."

kimchy replied:
will change

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java:
```diff
@@ -141,6 +148,89 @@ public String getDetails() {
+        for (ShardRouting shard : state.routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)) {
+            if (shard.primary() == false) {
+                IndexMetaData indexMetaData = state.metaData().index(shard.getIndex());
+                long delayTimeoutSetting = shard.unassignedInfo().getAllocationDelayTimeoutSetting(settings, indexMetaData.getSettings());
```
This is redundant. getDelayAllocationExpirationIn also calls getAllocationDelayTimeoutSetting()==0 and returns 0 in that case.

kimchy replied:
yea, agreed

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -173,6 +176,10 @@ public int getNumberOfInFlightFetch() {
         return this.numberOfInFlightFetch;
     }
 
+    public int getDelayedUnassignedShards() {
```
can we have a javadoc string for this?

kimchy replied:
sure

s1monw commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java:
```diff
@@ -62,6 +65,10 @@
+    private AtomicBoolean rerouting = new AtomicBoolean();
+
+    private volatile long registeredNextDelaySetting = Long.MAX_VALUE;
+    private volatile ScheduledFuture registeredNextDelayFuture;
```
this is updated in the cluster state update thread right? I think it doesn't need to be volatile then?

kimchy replied:
its sneaky, conceptually, it doesn't, due to the fact that we have single threaded cluster service execution, I think its more readable that way, so we don't worry about it

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java:
```diff
@@ -141,6 +148,89 @@ public String getDetails() {
+    /**
+     * Finds the next (closest) delay expiration setting of an unassigned shard. Returns 0 if there are none.
+     */
+    public static long findNextDelayedAllocationSetting(Settings settings, ClusterState state) {
```
Minor: can we call this findSmallestDelayedAllocationSettings? Next confused me implying it somehow depends on now.

kimchy replied:
sure

s1monw commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java:
```diff
@@ -139,9 +166,14 @@ private void reroute() {
             if (lifecycle.stopped()) {
                 return;
             }
+            if (rerouting.compareAndSet(false, true) == false) {
```
I think it would be clearer if we acquire a lock here somehow using `tryLock()`  this would make it clear that there is only one reroute happening at the same time?

kimchy replied:
its the same thing we do in the GatewayAllocator, I find this one simpler to reason about be honest

s1monw commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java:
```diff
@@ -139,9 +166,14 @@ private void reroute() {
+                logger.trace("already has pending reroute, ignoring");
+                return;
+            }
             clusterService.submitStateUpdateTask(CLUSTER_UPDATE_TASK_SOURCE, Priority.HIGH, new ClusterStateUpdateTask() {
```
can `clusterService.submitStateUpdateTask` throw an exception that prevents the reroute lock from being released?

kimchy replied:
I will protected against it, good point, I will also add it to the gateway allocator where we do something similar

bleskes commented on core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -509,6 +513,19 @@ public int compare(DiscoveryNode o1, DiscoveryNode o2) {
+                long delay = shard.unassignedInfo().getDelayAllocationExpirationIn(settings, indexMetaData.getSettings());
+                if (delay > 0) {
+                    logger.debug("[{}][{}]: delaying allocation of [{}] for [{}]", shard.index(), shard.id(), shard, TimeValue.timeValueMillis(delay));
+                    changed = true; // mark it as changed, since we want to kick a publishing to schedule future allocation
```
can we add a link to RoutingService#clusterChanged to the comment?

kimchy replied:
will do

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java:
```diff
@@ -125,6 +132,26 @@ public void clusterChanged(ClusterChangedEvent event) {
                 }
             }
+
+            long nextDelaySetting = UnassignedInfo.findNextDelayedAllocationSetting(settings, event.state());
```
as per discussion can we add a small comment about the rational here? (figure out when the next unassigned allocation need to happen from now. If this is larger or equal then the last time we checked and scheduled, we are guaranteed to have a reroute until then, so no need to schedule again).

kimchy replied:
sure

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java:
```diff
@@ -125,6 +132,26 @@ public void clusterChanged(ClusterChangedEvent event) {
+                        logger.warn("failed to schedule/execute reroute post unassigned shard", t);
+                    }
+                });
+            }
```
can we add a trace log on as an else so we know why we skipped scheduling if we need to?

s1monw commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java:
```diff
@@ -62,6 +64,10 @@
     private volatile boolean routingTableDirty = false;
 
     private volatile Future scheduledRoutingTableFuture;
```
as far as I can tell we are scheduleing on a fixed delay every 10 seconds so I wonder if we should just use this and don't schedule another task? I might be missing something

kimchy replied:
we don't do reroute every 10 seconds, we only do it if its dirty within this class

s1monw replied:
ok but we schedule at fixed delay so can't we just check if we are dirty due to a delayed allocation? I mean that would just be 10 sec. off in that case and we can skip the second scheduled task?

kimchy replied:
I don't think its a large overhead? it makes things simpler if someone sets it to a small value than 10s when they test and such?

s1monw replied:
well I think it makes the code much simper if we only ahve one... for tests we can just decrease the interval for the scheduled task?

kimchy replied:
to be honest, I really want to get away from this scheduling completely, I started to try and do it in this change, and it became bigger and non relevant (reroute after node added in disco). My goal is to have a single place in this class to handle calling reroute, use it in allocator and shard failed one, and remove completely the scheduled nature of it, just prompt reroute. For this reason, I prefer not to rely on this time base logic, and I think its nicely encapsulated into the delay nature of when we reroute.

kimchy commented:
@s1monw @bleskes pushed another set of changes

bleskes commented:
LGTM +1

s1monw commented:
LGTM makes sense @kimchy

kimchy commented:
pushed to master and 1.x, @clintongormley I forgot to add the docs, where do you think it makes sense to document this?

clintongormley commented:
@kimchy i'd say in the Index Shard Allocation page: https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules-allocation.html

plus a note on the cluster health page

bleskes commented:
I think it’s also good to mention on the (rolling) upgrade docs: docs/reference/setup/upgrade.asciidoc


> On 18 Jun 2015, at 21:47, Clinton Gormley <notifications@github.com> wrote:
> 
> @kimchy i'd say in the Index Shard Allocation page: https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules-allocation.html
> 
> plus a note on the cluster health page
> 
> —
> Reply to this email directly or view it on GitHub.
>
